Documentation for := vs <> (#2222)
chipsalliance/chisel3: docs/src/explanations/Documentation-of-functions.md

```
## Concept 1: <> is Communicative

This experiment is set up to test for the function of `<>` using the experiment above.

Achieving this involves flipping the RHS and LHS of each arrow and seeing how '<>'  will react.
( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  io.in := p.io.a
  // connect producer to consumer
  p.io.b := c.io.a
  // connect consumer to IO
  c.io.b := io.out
}

class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion: 
The Verilog remained the same without incurring errors, showing that the `<>` operator is generally communicative.

> 

Reviewer: I'd like to add another concept...

### `<>` and `:=` connects signals by field name

Scastie Link: https://scastie.scala-lang.org/Uf4tQquvQYigZAW705NFIQ

This experiment creates a `MockDecoupledIO` which has the same fields by name as a `DecoupledIO`. Chisel lets us connect it and produces the same verilog, even though `MockDecoupledIO` and `DecoupledIO` are different tyeps.

And here is another experiment, where we remove one of the fields of MockDecoupledIO:
https://scastie.scala-lang.org/ChtkhKCpS9CvJkjjqpdeIA
This one fails because there is something missing.
It doesn't matter the order -- they need to be the same.
Same is true for `:=`... the Scala types do not need to match but all the signals must match up... but may want to run a few more experiments here to make sure whether LHS and RHS must have exactly the same signals, or all signals from LHS must be present in RHS, or vice versa

Reviewer: This is what you are showing in Concept 7 now, so I will mark this one as resolved

## Concept 2: := means assign ALL signals from the RHS to the LHS, regardless of their direction.
Using the same experiment code as above, we set to test for the function of ":="
To achieve this, "<>" is being replaced with ":=" in the sample code above.
(Scastie link to the experiment:https://scastie.scala-lang.org/Shorla/o1ShdaY3RWKf0IIFwwQ1UQ)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  p.io.a <> io.in
  // connect producer to consumer
  c.io.b <> p.io.b
  // connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting error message for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:

> Trying to use ":=" to connect an input will give this error, whether flipped or not.

Reviewer: ```suggestion
The `:=` operator goes field-by-field and attempts to connect the RHS to the LHS. If something on the LHS is actually an `Input`, or something on the RHS is an `Output`, you will get this error:
```

## Concept 3: Always Use := to assign DontCare to Wires

> ":=" or "<>" which is best used to assign wires of unknown direction to DontCare? 
> We will find that out using the sample codes above.
> ( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

Reviewer: ```suggestion

When assigning 1DontCare1 to something that is not directioned, should you use `:=` or `<>`? which is best used to assign wires of unknown direction to DontCare? 
We will find out using the sample codes below:
( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)
```

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp := DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:

> When '<>' was used to assign a wire with an unknown direction "tmp" to DontCare, the following error message appeared.
> But when ":=" was used to assign the wire to DontCare, No errors came up. 
> 

Reviewer: ```suggestion
When `<>` was used to assign the unidrectioned wire `tmp` to DontCare, we got an error.
But when `:=` was used to assign the wire to DontCare, no errors will occur.

Thus, when assigning DontCare to a wire, always use `:=`.

```

##  Concept 4: You can use <> or := to assign DontCare to directioned things (IOs)

> ':=' or '<>' which is best used to assign directioned things to DontCare? 
> We will find that out using the sample codes above.
> ( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

Reviewer: ```suggestion
When assigning `DontCare` to something that is directioned, should you use `:=` or `<>`? 
We will find out using the sample codes below:
( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)
```

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp <> DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```

> ### Conclusion: 
> Both "<>" and ":=" can be used to assign directioned things (IOs) to DontCare as shown in "io.in" and "p.io.a" respectively. This is basically equivalent because in this case both "<>" and ":=" will figure the direction from the LHS.
> 

Reviewer: ```suggestion
### Conclusion: 
Both `<>` and `:=` can be used to assign directioned things (IOs) to DontCare as shown in `io.in` and `p.io.a` respectively. This is basically equivalent because in this case both `<>` and `:=` will determine the direction from the LHS.

```

I'm really not sure about this explanation. I don't know if this is why it works

> ##  Concept 5:  `<>` Cannot be used to connect two non-directioned things (Wires)
> The goal is to check if <> can connect two wires using the Experiment code above.
> ( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

Reviewer: ```suggestion
##  Concept 5:  `<>` Cannot be used to connect two non-directioned things (Wires)
The goal is to check if `<>` can connect two `Wire`s using the Experiment code above.
( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)
```

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp := DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:

> The code above shows that <> can't connect two wires, this is because Chisel can't figure out which way things flow. If it is used this is the expected error.
> 

Reviewer: ```suggestion
The code above shows that `<>` can't connect two `Wire`s, this is because Chisel can't figure out which way things flow. If it is used this is the expected error.

```

### Experiment Setup

```scala mdoc
// Imports used by the following examples
import chisel3._
import chisel3.util.DecoupledIO
```

The diagram for the experiment can be viewed **[here](https://docs.google.com/document/d/14C918Hdahk2xOGSJJBT-ZVqAx99_hg3JQIq-vaaifQU/edit?usp=sharing)**
![Experiment Image](https://raw.githubusercontent.com/chipsalliance/chisel3/master/docs/src/images/chisel_01.png?sanitize=true)

```scala mdoc:silent

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  // connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```

> ## Concept 1: <> is Communicative

Reviewer 2: ```suggestion
## Concept 1: <> is Commutative
```

This experiment is set up to test for the function of `<>` using the experiment above.

Achieving this involves flipping the RHS and LHS of each arrow and seeing how '<>'  will react.
( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  io.in := p.io.a
  // connect producer to consumer
  p.io.b := c.io.a
  // connect consumer to IO
  c.io.b := io.out
}

class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })

>   io.b <> io.a

Reviewer 2: This is the same module as above... To demonstrate commutativity, we want to swap the arguments:

```suggestion
  io.a <> io.b
```

}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion: 

> The Verilog remained the same without incurring errors, showing that the `<>` operator is generally communicative.

Reviewer 2: ```suggestion
The Verilog remained the same without incurring errors, showing that the `<>` operator is generally commutative.
```

> ## Concept 2: := means assign ALL signals from the RHS to the LHS, regardless of their direction.

Reviewer 2: This is not true. The RHS can have more `Record` fields than the LHS; in this case, only the RHS elements that can be mapped to LHS are assigned.

Author: @debs-sifive so not all signals From the RHS are assigned to the LHS, Does the experiment on concept 2 show that?

Reviewer 2: No, since you're connecting two of the same kind of bundle. The RHS bundle can have more elements in it than the LHS.

Reviewer: I think we need a different, simpler experiment that just shows this for single-directional bundle, where one side has an `a`, `b`, and `c` and the other only has `a`, `b`.... and show what happens when they are on either side of the `:=`

Author: Okay, but will be new experiment still be in this document or an entire new document will have to be created for that?

Reviewer: > The RHS bundle can have more elements in it than the LHS.

@debs-sifive that is not what my experimenting shows in https://scastie.scala-lang.org/MOR2DI4KT9KxElvXBWYiWA ... the error is `chisel3.internal.ChiselException: Connection between left (FilterIO(IO in unelaborated Block)) and source (NotReallyAFilterIO(IO in unelaborated Block)) failed @.z: Left Record missing field (z).`

Are you sure about the statement? Or is what you are saying true only for `:=`, not `<>`?

Reviewer: Confirmed that it gives similar error for `:=`

Reviewer: > Okay, but will the new experiment still be in this document or an entirely new document will have to be created for that?

@Shorla  the simpler document is in the existing code about "Interfaces and Connections" that we discussed yesterday

Reviewer: There is some confusion in this comment thread between this concept 2 and this concept 7. This is about ignoring direction, where concept 7 is about signals being present or not. Let's keep this example about the directionality aspect.

Reviewer: Coercing inputs to inputs does not work: https://scastie.scala-lang.org/reYHtvBMS32xK4O6mj7xLg

What about outputs... does work: https://scastie.scala-lang.org/rSRU7as8SQC7MG8vevzliQ
Need to check th eresult of that and see what it generated...

Reviewer 2: @mwachs5 my comment was just referring to the `:=` MonoConnect. I'm going off the documentation here: https://github.com/chipsalliance/chisel3/blob/849d4a0b7f6f7ea056c5280b9d319dadf5225022/core/src/main/scala/chisel3/internal/MonoConnect.scala#L22

This example demonstrates a compiling `:=` connection with more elements on the RHS than LHS: https://scastie.scala-lang.org/opTU5MznT5OmQhIIsYUylA

The `<>` BiConnect is expected to error if Bundle elements do not match up exactly, as documented here:
https://github.com/chipsalliance/chisel3/blob/849d4a0b7f6f7ea056c5280b9d319dadf5225022/core/src/main/scala/chisel3/internal/BiConnect.scala#L23

Reviewer: Ok. I am realizing that a lot of confusion is because this section is called "Bulk Connect". but `:=` is a bulk connect just as much as `<>` is. I think we should rename this header from "Bulk Conections" to "Bidirectional Connections `<>`"

Using the same experiment code as above, we set to test for the function of ":="
To achieve this, "<>" is being replaced with ":=" in the sample code above.
(Scastie link to the experiment:https://scastie.scala-lang.org/Shorla/o1ShdaY3RWKf0IIFwwQ1UQ)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  p.io.a <> io.in
  // connect producer to consumer
  c.io.b <> p.io.b
  // connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })

>   io.b <> io.a

Reviewer 2: Did you mean the following? Otherwise you are still using the `BiConnect` (`<>`) from the previous example.

```suggestion
  io.b := io.a
```

}
```
Below we can see the resulting error message for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:
The `:=` operator goes field-by-field and attempts to connect the RHS to the LHS. If something on the LHS is actually an `Input`, or something on the RHS is an `Output`, you will get this error:

> 

Reviewer 2: There's no error message here.

Author: Hi @debs-sifive are you requesting I add the error message to the document? I only just removed it.

Reviewer 2: Ok, could you change this wording or delete this paragraph then? It implies that there will be an example error message beneath it.

Author: Okay noted.

## Concept 3: Always Use := to assign DontCare to Wires

When assigning `DontCare` to something that is not directioned, should you use `:=` or `<>`? 
We will find out using the sample codes below:
( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp := DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a

> }

Reviewer 2: This example can be simplified a lot. I don't think you need most of the circuitry here, besides what is demonstrating the `DontCare`.

```
Below we can see the resulting Verilog for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:
When `<>` was used to assign the unidrectioned wire `tmp` to DontCare, we got an error.
But when `:=` was used to assign the wire to DontCare, no errors will occur.

Thus, when assigning DontCare to a wire, always use `:=`.

##  Concept 4: You can use <> or := to assign DontCare to directioned things (IOs)
When assigning `DontCare` to something that is directioned, should you use `:=` or `<>`? 
We will find out using the sample codes below:
( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))

>   tmp <> DontCare

Reviewer 2: I don't think you need more in this example than these lines.

Author: Noted

---

layout: docs

title:  "Documention of functions"

section: "chisel3"

---

> # Documentation of Functions := vs <>

Reviewer: ```suggestion
# Further Explanation for the := and  <> Operators
```

> The difference in the function of := and <> has been a major source of debate in lots of conversations
> 
> This document explains the difference by setting up an experiment using Scastie examples + DecoupledIO examples to put the functions to the test.

Reviewer: ```suggestion
Chisel contains two connection operators, `:=` and `<>`. This document serves for a deeper explanation of the differences of the two and when to use one or the other. The differences are demonstrated with experiments using Scastie examples + DecoupledIO examples.
```

> ### Experiment

Reviewer: ```suggestion
### Experiment Setup
```

```scala mdoc
// Imports used by the following examples
import chisel3._
import chisel3.util.DecoupledIO
```

> The diagram for the experiment can be viewed **[here](https://docs.google.com/document/d/14C918Hdahk2xOGSJJBT-ZVqAx99_hg3JQIq-vaaifQU/edit?usp=sharing)**

Reviewer: we should figure out how to embed the actual diagram... we'll need to read the markdown docs

Author: I will go through that and see how I can put it in

( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  io.in <> p.io.a
  // connect producer to consumer
  p.io.b <> c.io.a
  // connect consumer to IO
  c.io.b <> io.out
}

class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion: 
The Verilog remained the same without incurring errors, showing that the `<>` operator is generally communicative.

## Concept 2: := means assign ALL signals from the RHS to the LHS, regardless of their direction.
Using the same experiment code as above, we set to test for the function of ":="

> To achieve this, "<>" is being replaced with ":=" in the sample code above and the RHS and LHS are flipped.

Reviewer: ```suggestion
To achieve this, "<>" is being replaced with ":=" in the sample code above and the RHS and LHS are flipped.
```

Why are we also doing the RHS and LHS flipping in this experiment? I think it is conflating two different concepts. Shouldn't we just do an experiment where we start from the original and replace `<>` with `:=`?

Author: I corrected that now

(Scastie link to the experiment:https://scastie.scala-lang.org/Shorla/o1ShdaY3RWKf0IIFwwQ1UQ)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  io.in := p.io.a
  // connect producer to consumer
  p.io.b := c.io.a
  // connect consumer to IO
  c.io.b := io.out
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```

> Below we can see the resulting Verilog for this example:
> ```scala modc:crash
> ChiselStage.emitVerilog(new Wrapper)

Reviewer: ```suggestion
Below we can see the resulting error message for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```

```
### Conclusion:

> Trying to use ":=" to connect an input will give this error, whether flipped or not.

Reviewer: Do we need to show the error again? isn't it sown in the `mdoc:crash` output?

Author: I have also removed error messages.

( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp <> DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion: 
Both "<>" and ":=" can be used to assign directioned things (IOs) to DontCare as shown in "io.in" and "p.io.a" respectively. This is basically equivalent because in this case both "<>" and ":=" will figure the direction from the LHS.

> ##  Concept 5: Use <> for intermediate wire connections

Reviewer: ```suggestion
##  Concept 5:  `<>` Cannot be used to connect two non-directioned things (Wires)
```

  //connect Producer to IO
  io.in := DontCare
  p.io.a <> DontCare
  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))
  tmp := DontCare
  p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:
The code above shows that <> can't connect two wires, this is because Chisel can't figure out which way things flow. If it is used this is the expected error.

~~~
chisel3.internal.ChiselException: Connection between left (DecoupledIO(Wire in Wrapper)) and source (DecoupledIO(Wire in Wrapper)) failed @.bits: Locally unclear whether Left or Right (both internal)
	at ... ()
	at Playground$Wrapper.<init>(main.scala:22)
	at Playground$$anonfun$17.apply(main.scala:40)
	at Playground$$anonfun$17.apply(main.scala:40)
	at ... ()
	at ... (Stack trace trimmed to user code only. Rerun with --full-stacktrace to see the full stack trace)
~~~

> ## Concept 6: <> only works between things with at least one known flow (An IO or child's IO). 

Reviewer: ```suggestion
## Concept 6: `<>`  works between things with at least one known flow (An IO or child's IO). 
```

If there is at least one known flow what will <> do? This will be showed using the experiment code.
( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/gKx9ReLVTTqDTk9vmw5ozg)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  //connect Producer to IO
    // For this experiment, we add a temporary wire and see if it works...
  //p.io.a <> io.in
  val tmp = Wire(DecoupledIO(UInt(8.W)))
  // connect intermediate wire
  tmp <> io.in
  p.io.a <> tmp
  //p.io.a <> io.in
  // connect producer to consumer
  c.io.a <> p.io.b
  //connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b <> io.a
}
```
Below we can see the resulting Verilog for this example:
```scala modc
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:

> The connection above went smoothly with no errors, this goes to show <> will work if there is only one wire in between because the IO and the childIO "fix" the direction.

Reviewer: ```suggestion
The connection above went smoothly with no errors, this goes to show <> will work as long as there is at least one directioned thing (IO or submodule's IO) to "fix" the direction.
```

> ---

Reviewer: let's rename this file to something like connection-operators.md

layout: docs

> title:  "Documention of functions"

Reviewer: ```suggestion
title:  "Deep Dive into Connection Operators"
```

section: "chisel3"

---

# Further Explanation for the := and  <> Operators

> Chisel contains two connection operators, `:=` and `<>`. This document serves for a deeper explanation of the differences of the two and when to use one or the other. The differences are demonstrated with experiments using Scastie examples + DecoupledIO examples.

Reviewer: ```suggestion
Chisel contains two connection operators, `:=` and `<>`. This document provides a deeper explanation of the differences of the two and when to use one or the other. The differences are demonstrated with experiments using Scastie examples + DecoupledIO examples.
```

Using the same experiment code as above, we set to test for the function of ":="
To achieve this, "<>" is being replaced with ":=" in the sample code above.
(Scastie link to the experiment:https://scastie.scala-lang.org/Shorla/o1ShdaY3RWKf0IIFwwQ1UQ)

```scala mdoc:silent:reset
import chisel3._
import chisel3.util.DecoupledIO

class Wrapper extends Module{
  val io = IO(new Bundle {
  val in = Flipped(DecoupledIO(UInt(8.W)))
  val out = DecoupledIO(UInt(8.W))
  })
  val p = Module(new PipelineStage)
  val c = Module(new PipelineStage) 
  // connect Producer to IO
  p.io.a <> io.in
  // connect producer to consumer
  c.io.b <> p.io.b
  // connect consumer to IO
  io.out <> c.io.b
}
class PipelineStage extends Module{
  val io = IO(new Bundle{
    val a = Flipped(DecoupledIO(UInt(8.W)))
    val b = DecoupledIO(UInt(8.W))
  })
  io.b := io.a
}
```
Below we can see the resulting error message for this example:
```scala modc:crash
ChiselStage.emitVerilog(new Wrapper)
```
### Conclusion:
The `:=` operator goes field-by-field and attempts to connect the RHS to the LHS. If something on the LHS is actually an `Input`, or something on the RHS is an `Output`, you will get this error:

> 

Reviewer: we are no longer showing the error here, what is the error we get?

Author: This is the error we got:
chisel3.internal.ChiselException: Connection between sink (DecoupledIO(IO io_a in PipelineStage)) and source (DecoupledIO(IO in unelaborated Wrapper)) failed @.ready: Sink is unwriteable by current module.
	at ... ()
	at Playground$Wrapper.<init>(main.scala:15)
	at Playground$$anonfun$13.apply(main.scala:31)
	at Playground$$anonfun$13.apply(main.scala:31)
	at ... ()
	at ... (Stack trace trimmed to user code only. Rerun with --full-stacktrace to see the full stack trace)

Author: I removed the errors because I did the same for other examples with errors.
Should I add the error or Just rewrite the conclusion, so it doesn't require adding the error message?

Reviewer: I think you can include the mdoc-generated error message below, or say something like "the error message shown above"... as it was written it seemed like the error would be shown below.

This document explains the difference by setting up an experiment using Scastie examples + DecoupledIO examples to put the functions to the test.

### Experiment

```scala mdoc:silent

// Imports used by the following examples

import chisel3._

import chisel3.util.DecoupledIO

```

The diagram for the experiment can be viewed **[here](https://docs.google.com/document/d/14C918Hdahk2xOGSJJBT-ZVqAx99_hg3JQIq-vaaifQU/edit?usp=sharing)**

```scala mdoc:silent

class Wrapper extends Module{

  val io = IO(new Bundle {

  val in = Flipped(DecoupledIO(UInt(8.W)))

  val out = DecoupledIO(UInt(8.W))

  })

  val p = Module(new PipelineStage)

  val c = Module(new PipelineStage) 

  // connect Producer to IO

  p.io.a <> io.in

  // connect producer to consumer

  c.io.a <> p.io.b

  // connect consumer to IO

  io.out <> c.io.b

}

class PipelineStage extends Module{

  val io = IO(new Bundle{

    val a = Flipped(DecoupledIO(UInt(8.W)))

    val b = DecoupledIO(UInt(8.W))

  })

  io.b <> io.a

}

```

Below we can see the resulting Verilog for this example:

```scala modc

ChiselStage.emitVerilog(new Wrapper)

```

## Concept 1: <> is Communicative

> This experiment is set up to test for the function of "<>" using the experiment above.

Reviewer: ```suggestion
This experiment is set up to test for the function of `<>` using the experiment above.
```

> Achieving this involves flipping the RHS and LHS of each arrow and seeing how "<>"  will react.

Reviewer: ```suggestion
Achieving this involves flipping the RHS and LHS of each arrow and seeing how `<>`  will react.
```

> ( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )

Reviewer: ```suggestion
( Scastie link for the experiment:https://scastie.scala-lang.org/Shorla/LVhlbkFQQnq7X3trHfgZZQ )
```

> ```scala mdoc:silent

Reviewer: does this work? I imagine you might need to have `scala mdoc:silent:reset` otherwise you would get an error that Wrapper was already defined?

class Wrapper extends Module{

  val io = IO(new Bundle {

  val in = Flipped(DecoupledIO(UInt(8.W)))

  val out = DecoupledIO(UInt(8.W))

  })

  val p = Module(new PipelineStage)

  val c = Module(new PipelineStage) 

  // connect Producer to IO

  io.in <> p.io.a

  // connect producer to consumer

  p.io.b <> c.io.a

  // connect consumer to IO

  c.io.b <> io.out

}

class PipelineStage extends Module{

  val io = IO(new Bundle{

    val a = Flipped(DecoupledIO(UInt(8.W)))

    val b = DecoupledIO(UInt(8.W))

  })

  io.b <> io.a

}

```

> Below we can see the resulting Verilog for this example:

Reviewer: ```suggestion
Below we can see the resulting Verilog for this example. Note that it is identical to the original verilog, showing that the `<>` operator is communicative:
```

```scala modc

ChiselStage.emitVerilog(new Wrapper)

```

### Conclusion: 

> The Verilog remained the same without incurring errors.

Reviewer: ```suggestion
The Verilog remained the same without incurring errors, showing that the `<>` operator is generally communicative.
```

## Concept 2: := means assign ALL signals from the RHS to the LHS, regardless of their direction.

Using the same experiment code as above, we set to test for the function of ":="

To achieve this, "<>" is being replaced with ":=" in the sample code above and the RHS and LHS are flipped.

(Scastie link to the experiment:https://scastie.scala-lang.org/Shorla/o1ShdaY3RWKf0IIFwwQ1UQ)

```scala mdoc:silent

class Wrapper extends Module{

  val io = IO(new Bundle {

  val in = Flipped(DecoupledIO(UInt(8.W)))

  val out = DecoupledIO(UInt(8.W))

  })

  val p = Module(new PipelineStage)

  val c = Module(new PipelineStage) 

  // connect Producer to IO

  io.in := p.io.a

  // connect producer to consumer

  p.io.b := c.io.a

  // connect consumer to IO

  c.io.b := io.out

}

class PipelineStage extends Module{

  val io = IO(new Bundle{

    val a = Flipped(DecoupledIO(UInt(8.W)))

    val b = DecoupledIO(UInt(8.W))

  })

  io.b <> io.a

}

```

Below we can see the resulting Verilog for this example:

> ```scala modc

Reviewer: ```suggestion
```scala mdoc:crash
```

I think we need that `crash` to show the error message otherwise building the docs will fail

	at ... ()

	at ... (Stack trace trimmed to user code only. Rerun with --full-stacktrace to see the full stack trace)

~~~

## Concept 3: Always Use := to assign DontCare to Wires

":=" or "<>" which is best used to assign wires of unknown direction to DontCare? 

We will find that out using the sample codes above.

( Scastie link for the ecperiment:https://scastie.scala-lang.org/Shorla/ZIGsWcylRqKJhZCkKWlSIA/1)

```scala mdoc:silent

class Wrapper extends Module{

  val io = IO(new Bundle {

  val in = Flipped(DecoupledIO(UInt(8.W)))

  val out = DecoupledIO(UInt(8.W))

  })

  val p = Module(new PipelineStage)

  val c = Module(new PipelineStage) 

  //connect Producer to IO

  io.in <> DontCare

  p.io.a <> DontCare

  val tmp = Wire(Flipped(DecoupledIO(UInt(8.W))))

  tmp <> DontCare

  p.io.a <> io.in

  // connect producer to consumer

  c.io.a <> p.io.b

  //connect consumer to IO

  io.out <> c.io.b

}

class PipelineStage extends Module{

  val io = IO(new Bundle{

    val a = Flipped(DecoupledIO(UInt(8.W)))

    val b = DecoupledIO(UInt(8.W))

  })

  io.b <> io.a

}

```

Below we can see the resulting Verilog for this example:

> ```scala modc

Reviewer: ```suggestion
```scala mdoc:crash
```

again you can combine the "here is the error message you will see" code with the `mdoc:crash` flag
